### PR TITLE
Add service-name check and regress test

### PR DESCRIPTION
**Pull request by yosuke-wolfssl**
This PR adds the service-name check to comply with RFC 4252 section 5.
If service name is not equal to "ssh-connection", SSH_MSG_USERAUTH_FAILURE would be sent to the peer.
Also, new unit test is added to exercise service-name validation.

### CODE REVIEW

File: tests/unit.c [hunk 1]
```diff
@@ -918,6 +918,86 @@ static int test_DoChannelRequest(void)
...
+        /* service name */
+        buf[len++] = (byte)(snsz >> 24); buf[len++] = (byte)(snsz >> 16);
+        buf[len++] = (byte)(snsz >>  8); buf[len++] = (byte)snsz;
+        WMEMCPY(buf + len, cases[i].svcName, snsz); len += snsz;
+
+        /* auth method: "none" */
+        buf[len++] = 0; buf[len++] = 0; buf[len++] = 0; buf[len++] = 4;
+        WMEMCPY(buf + len, "none", 4); len += 4;
+
+        ret = wolfSSH_TestDoUserAuthRequest(ssh, buf, len, &idx);
```
[the remark this answers is not in the document]
Comment (yosuke-wolfssl) [reply tests/unit.c]: Fixed

File: tests/unit.c [hunk 2]
```diff
@@ -918,6 +918,86 @@ static int test_DoChannelRequest(void)
...
+
+        /* service name */
+        buf[len++] = (byte)(snsz >> 24); buf[len++] = (byte)(snsz >> 16);
+        buf[len++] = (byte)(snsz >>  8); buf[len++] = (byte)snsz;
+        WMEMCPY(buf + len, cases[i].svcName, snsz); len += snsz;
+
+        /* auth method: "none" */
+        buf[len++] = 0; buf[len++] = 0; buf[len++] = 0; buf[len++] = 4;
+        WMEMCPY(buf + len, "none", 4); len += 4;
+
```
[the remark this answers is not in the document]
Comment (yosuke-wolfssl) [reply tests/unit.c]: Fixed

File: src/internal.c [hunk 3]
```diff
@@ -8401,10 +8402,19 @@ static int DoUserAuthRequest(WOLFSSH* ssh,
         authData.serviceName = buf + begin;
         begin += authData.serviceNameSz;
 
-        ret = GetSize(&authData.authNameSz, buf, len, &begin);
+        if (NameToId((const char*)authData.serviceName, authData.serviceNameSz)
+                != ID_SERVICE_CONNECTION) {
+            WLOG(WS_LOG_DEBUG, "DUAR: Invalid service name");
```
Comment (padelsbach) [inline src/internal.c:8372]: The other debug logs in this function do not have `DUAR` -- consider updating those to match this one, following convention from `DoUserAuthRequestNone`.

File: src/internal.c [hunk 4]
```diff
@@ -8401,10 +8402,19 @@ static int DoUserAuthRequest(WOLFSSH* ssh,
...
+            serviceValid = 0;
+            ret = SendUserAuthFailure(ssh, 0);
+            *idx = len;
```
Comment (padelsbach) [inline src/internal.c:8376]: Other error cases in this function do not set `*idx`. Why is this needed for this specific error? Can you add a comment?

**Comment by padelsbach** [comment]
@yosuke-wolfssl, can you please rebase to resolve conflicts?

File: src/internal.c [hunk 5]
```diff
@@ -8401,10 +8402,19 @@ static int DoUserAuthRequest(WOLFSSH* ssh,
         authData.serviceName = buf + begin;
         begin += authData.serviceNameSz;
 
-        ret = GetSize(&authData.authNameSz, buf, len, &begin);
+        if (NameToId((const char*)authData.serviceName, authData.serviceNameSz)
```
Comment (aidangarske) [inline src/internal.c:8370]: **🔵 [Low] Inconsistent cast style in new NameToId call**

The new service-name check casts `authData.serviceName` as `(const char*)`, but the immediately adjacent NameToId call on line 8420 for `authData.authName` uses `(char*)`. Both compile fine because `NameToId` takes `const char*`, but the mixed styles inside the same function are distracting.

**Fix:** Pick one cast style for NameToId in this function (either both `(char*)` or both `(const char*)`) so the two call sites read identically.

**Review by aidangarske — COMMENTED** [review]
## Skoll Code Review

**Scan type:** `review`
**Overall recommendation:** `APPROVE`
**Findings:** 7 total — 7 posted, 0 skipped
**7 finding(s) posted as inline comments** (see file-level comments below)

### Posted findings
- [Low] **Inconsistent cast style in new NameToId call** — `src/internal.c:8405-8406`
- [Low] **Error from SendUserAuthFailure overwrites ret, making invalid-service path return non-success** — `src/internal.c:8405-8411`
- [Low] **`*idx = len` is written even when SendUserAuthFailure returned an error** — `src/internal.c:8409-8411`
- [Low] **Test assumes output path succeeds on a fresh session without documentation** — `tests/unit.c:952-1041`
- [Low] **Test does not cover empty or oversize service names** — `tests/unit.c:964-973`
- [Info] **`serviceValid` could be replaced by a guard-style early return block for readability** — `src/internal.c:8370,8417`
- [Info] **`*idx = len` set before SendUserAuthFailure result is checked** — `src/internal.c:8409-8410`

*Review generated by [Skoll](https://github.com/wolfSSL/fenrir/tree/main/skoll)*

File: src/internal.c (hunk 5 above)
Comment (aidangarske) [inline src/internal.c:8370]: **🔵 [Low] Error from SendUserAuthFailure overwrites ret, making invalid-service path return non-success**

When the service name is invalid the code calls `ret = SendUserAuthFailure(ssh, 0)`. If `SendUserAuthFailure()` fails (e.g., `WS_SOCKET_ERROR_E`, `WS_WANT_WRITE`), that error is returned to the outer dispatcher as if parsing itself failed. In the unknown-auth-method path below (line 8444) the same pattern is used, so this is internally consistent with the existing convention — but it is worth confirming that propagating a send error here is the intended behavior (e.g., `WS_WANT_WRITE` from a non-blocking socket could reach `DoReceive` here and bubble up differently than the caller expects). The unit test uses a capture-callback that always returns success, so this path is not exercised.

**Fix:** Confirm that it is acceptable to return the `SendUserAuthFailure()` result (including `WS_WANT_WRITE`) to the `DoReceive` loop when the only parsing failure was a bad service name. If the intent is "we handled this request, stay alive," consider saving the send result separately from `ret` or adding a test where the IO callback returns `WS_CBIO_ERR_WANT_WRITE`.

File: src/internal.c [hunk 6]
```diff
@@ -8401,10 +8402,19 @@ static int DoUserAuthRequest(WOLFSSH* ssh,
...
+                != ID_SERVICE_CONNECTION) {
+            WLOG(WS_LOG_DEBUG, "DUAR: Invalid service name");
+            serviceValid = 0;
+            ret = SendUserAuthFailure(ssh, 0);
```
Comment (aidangarske) [inline src/internal.c:8374]: **🔵 [Low] `*idx = len` is written even when SendUserAuthFailure returned an error**

On the invalid-service path, `ret = SendUserAuthFailure(ssh, 0); *idx = len;` advances idx unconditionally. If `SendUserAuthFailure` returns a non-success value (e.g., a non-blocking `WS_WANT_WRITE` or an allocation failure inside `PreparePacket`), the function returns the error to `DoPacket` but idx is still fully consumed. For most error paths this is harmless because the caller tears the connection down, but it does mean the decoded-length cursor no longer matches what was actually processed. Other handlers in this file leave idx alone when ret is non-success.

**Fix:** Consider only advancing `*idx = len` when `SendUserAuthFailure` returns `WS_SUCCESS`, matching the idx-preservation convention used elsewhere in this function on error paths.

File: tests/unit.c [hunk 7]
```diff
@@ -918,6 +918,128 @@ static int test_DoChannelRequest(void)
...
+ * check and still satisfy all three assertions — proving it never tries to
+ * parse the missing auth-method field.  If the short-circuit were absent,
+ * GetSize() for authNameSz would hit end-of-buffer and return WS_BUFFER_E,
+ * failing assertion 1.
+ *
+ * For the valid-service case, auth method "xyz-unknown" (always unsupported
+ * regardless of compile-time options) is included.  The function reaches
+ * auth-method dispatch, falls to the unknown-method else-branch, and sends
+ * USERAUTH_FAILURE via that normal path. */
+static int test_DoUserAuthRequest_serviceName(void)
```
Comment (aidangarske) [inline tests/unit.c:1376]: **🔵 [Low] Test assumes output path succeeds on a fresh session without documentation**

The test relies on no cipher being negotiated so the captured packet is plaintext with `MSGID_USERAUTH_FAILURE` at byte offset 5. That assumption is already documented in `test_DoChannelRequest` (lines 792-797) but not here. A one-line comment referencing the same packet-layout invariant would save the next reader from retracing the logic.

**Fix:** Add a short comment explaining why byte 5 holds the message ID, for parity with `test_DoChannelRequest`.

File: tests/unit.c [hunk 8]
```diff
@@ -918,6 +918,128 @@ static int test_DoChannelRequest(void)
...
+    WOLFSSH* ssh = NULL;
+    int result = 0;
+    struct {
+        const char* svcName;
+        word32      svcNameSz;
+        const char* authMethod;   /* NULL = omit field (proves short-circuit) */
+        word32      authMethodSz;
+        int         expectRet;
+        const char* label;
+    } cases[] = {
```
Comment (aidangarske) [inline tests/unit.c:1390]: **🔵 [Low] Test does not cover empty or oversize service names**

The new test exercises three cases (`ssh-connection`, `ssh-agent`, `bad`). It does not cover the boundary cases `serviceNameSz == 0` (zero-length service name) or a service name equal to another known service ID (e.g., `ssh-userauth`). Both should also be rejected by the new `NameToId(...) != ID_SERVICE_CONNECTION` check and would provide stronger evidence that only `ssh-connection` is accepted. `NameToId("", 0)` returns `ID_UNKNOWN`, and `NameToId("ssh-userauth", 12)` returns `ID_SERVICE_USERAUTH`, both of which should short-circuit — but neither is currently verified.

**Fix:** Add a zero-length service-name case and an `ssh-userauth` service-name case to the `cases[]` table to verify the check is strict and not dependent on `NameToId` returning `ID_UNKNOWN` specifically.

File: src/internal.c [hunk 9]
```diff
@@ -8367,6 +8367,7 @@ static int DoUserAuthRequest(WOLFSSH* ssh,
     word32 begin;
     int ret = WS_SUCCESS;
     byte authNameId;
+    byte serviceValid = 1;
```
Comment (aidangarske) [inline src/internal.c]: **⚪ [Info] `serviceValid` could be replaced by a guard-style early return block for readability**

A new local `byte serviceValid = 1;` is introduced solely to skip the auth-method dispatch block after the short-circuit. The function already uses the chained `if (ret == WS_SUCCESS)` pattern throughout. An alternative would be to leave `ret` as `WS_SUCCESS` (since the packet was handled) and skip the dispatch by leaving `authData.authName` zero-length and `authNameId` == `ID_UNKNOWN`, which already falls into the else-branch at line 8441. But that would still call `wolfSSH_SetUsernameRaw` a second time and `SendUserAuthFailure` a second time, which is undesirable. The `serviceValid` flag is fine as-is.

**Fix:** Keep as-is. Flag is clearer than refactoring around the existing chained-if pattern.

File: src/internal.c (hunk 6 above)
Comment (aidangarske) [inline src/internal.c:8374]: **⚪ [Info] `*idx = len` set before SendUserAuthFailure result is checked**

`*idx` is written unconditionally immediately after (and independent of) the `SendUserAuthFailure()` call. The surrounding convention in this function (and in `DoReceive`) is "advance the buffer index on success, leave untouched on failure." Here the index is advanced even when the send fails. The `DoReceive` dispatcher at src/internal.c:10178 will still advance `ssh->inputBuffer.idx` by `payloadIdx` even on error, so the behavior is not incorrect — but the asymmetry with the valid-path write `*idx = begin;` at line 8452 (guarded by `if (ret == WS_SUCCESS && serviceValid)`) reads slightly oddly.

**Fix:** Optional: move `*idx = len;` above the `SendUserAuthFailure` call to make it clear the assignment is unconditional and not contingent on send success, or leave as-is. Not a correctness issue.

File: src/internal.c (hunk 3 above)
Comment (yosuke-wolfssl) [reply src/internal.c:8372]: Fixed

File: src/internal.c (hunk 5 above)
Comment (yosuke-wolfssl) [reply src/internal.c:8370]: Fixed to cast both to const char*

File: src/internal.c (hunk 4 above)
Comment (yosuke-wolfssl) [reply src/internal.c:8376]: Added a comment

File: src/internal.c (hunk 6 above)
Comment (yosuke-wolfssl) [reply src/internal.c:8374]: This shall consume all of data to release the input buffer properly.
I fixed both of cases.

File: src/internal.c (hunk 9 above)
Comment (yosuke-wolfssl) [reply src/internal.c]: OK

File: src/internal.c (hunk 6 above)
Comment (yosuke-wolfssl) [reply src/internal.c:8374]: OK

File: src/internal.c (hunk 5 above)
Comment (yosuke-wolfssl) [reply src/internal.c:8370]: I'll keep current convention as-is.
I didn't find any problematic behaviour for now.

File: tests/unit.c (hunk 7 above)
Comment (yosuke-wolfssl) [reply tests/unit.c:1376]: Fixed

File: tests/unit.c (hunk 8 above)
Comment (yosuke-wolfssl) [reply tests/unit.c:1390]: Added two cases you mentioned.

**Comment by yosuke-wolfssl** [comment]
Hello @aidangarske ,

It's ready for review again.
Could you please take a look at this ?

File: src/internal.c [hunk 10]
```diff
@@ -8361,13 +8362,23 @@ static int DoUserAuthRequest(WOLFSSH* ssh,
         authData.serviceName = buf + begin;
         begin += authData.serviceNameSz;
 
-        ret = GetSize(&authData.authNameSz, buf, len, &begin);
```
Comment (ejohnstown) [inline src/internal.c:8383]: The function GetStringRef() was added for cases like this. It does all the appropriate length checking and it updates the idx. I just never got around to replacing all GetSize() and adjusting the idx myself with this function.

GetStringRef will give you a pointer into buf that points at the string, and it gives you the length. You don't need to free it, and the pointer is to const.

**Review by ejohnstown — CHANGES_REQUESTED** [review]

**Review by aidangarske — COMMENTED** [review]
## Skoll Code Review

**Scan type:** `review`
**Overall recommendation:** `COMMENT`
**Findings:** 3 total — 3 posted, 0 skipped
**3 finding(s) posted as inline comments** (see file-level comments below)

### Posted findings
- [Medium] **wolfSSH_SetUsernameRaw mutates session state before the service-name check** — `src/internal.c:8351-8376`
- [Low] **Unrelated blank-line removal between DoReceive and DoProtoId** — `src/internal.c:10719`
- [Low] **Test couples to internal packet layout (byte-5 message-id) without abstraction** — `tests/unit.c:1027-1035`

*Review generated by [Skoll](https://github.com/wolfSSL/fenrir/tree/main/skoll)*

File: src/internal.c [hunk 11]
```diff
@@ -8361,13 +8362,23 @@ static int DoUserAuthRequest(WOLFSSH* ssh,
...
+            WLOG(WS_LOG_DEBUG, "DUAR: Invalid service name");
+            serviceValid = 0;
+            ret = SendUserAuthFailure(ssh, 0);
+            /* Consume all remaining data */
+            *idx = len;
+        }
+        else {
+            ret = GetSize(&authData.authNameSz, buf, len, &begin);
+        }
     }
```
Comment (aidangarske) [inline src/internal.c:8382]: **🟠 [Medium] wolfSSH_SetUsernameRaw mutates session state before the service-name check**

The username is committed into `ssh->userName` at line 8351 *before* the new service-name validation runs. If the peer sends a request with a non-`ssh-connection` service name (or any other malformed thing later in parsing), the session keeps the attacker-supplied username while we send `USERAUTH_FAILURE`. The PR doesn't introduce this ordering, but by adding a new reject path between the username set and the auth-method dispatch it widens the window where this matters. A second retry will overwrite the field, so this is mostly cosmetic, but consider deferring `wolfSSH_SetUsernameRaw` until after we accept the request (or at least until after the service name is validated) so that rejected requests don't leave residual state. This would also let you drop the duplicate `wolfSSH_SetUsernameRaw` call at line 8410-8413.

**Fix:** Set the username only on a fully-validated request to avoid storing state from rejected attempts.

File: tests/unit.c [hunk 12]
```diff
@@ -918,6 +918,136 @@ static int test_DoChannelRequest(void)
...
+
+        if (ret != cases[i].expectRet) {
+            printf("DoUserAuthRequest_svcName[%s]: ret=%d expected=%d\n",
+                   cases[i].label, ret, cases[i].expectRet);
+            result = -502 - i;
+            break;
+        }
+
+        /* MSGID_USERAUTH_FAILURE must be in the captured packet. */
+        if (s_authSvcCaptureSz <= 5 ||
```
Comment (aidangarske) [inline tests/unit.c]: **🔵 [Low] Test couples to internal packet layout (byte-5 message-id) without abstraction**

The assertion `s_authSvcCapture[5] != MSGID_USERAUTH_FAILURE` hard-codes the offset of the message id in the framed packet (4-byte length + 1-byte padding length + msg id). This is the same approach used by `test_DoChannelRequest`, so the inconsistency is repo-wide rather than PR-introduced, but every new copy of this pattern increases the cleanup cost if/when packet framing changes. Consider extracting a small helper like `static int CaptureMsgId(byte* buf, word32 len)` that the existing channel-request test could also adopt.

**Fix:** Optional cleanup; not a merge blocker.

File: src/internal.c [hunk 13]
```diff
@@ -10716,7 +10729,6 @@ int DoReceive(WOLFSSH* ssh)
     return ret;
```
Comment (aidangarske) [inline src/internal.c:10734]: **🔵 [Low] Unrelated blank-line removal between DoReceive and DoProtoId**

The diff removes a blank line at the top of the gap between `DoReceive` and `DoProtoId`. This change has nothing to do with the service-name check and just adds noise to the diff. Consider keeping unrelated whitespace edits out of feature PRs to keep `git blame` clean.

**Fix:** Revert the stray formatting hunk.

---
> **Note:** Referenced line (`src/internal.c:10719`) is outside the diff hunk. Comment anchored to nearest changed region.

File: src/internal.c [hunk 14]
```diff
@@ -8390,8 +8401,10 @@ static int DoUserAuthRequest(WOLFSSH* ssh,
 #endif
         else {
```
Comment (aidangarske) [inline src/internal.c:8406]: **🔵 [Low] Unknown-auth-method `begin = len` consumption path lacks meaningful coverage** · _test-coverage_

The PR adds `begin = len;` (with comment `/* Consume all remaining data */`) in the unknown-auth-method else branch so that `DoUserAuthRequest()` defensively skips any trailing auth-method-specific payload when a request has a valid service name but an unsupported auth method. The new test `test_DoUserAuthRequest_serviceName` (tests/unit.c:953-1048) does exercise this branch via the `valid svc unknown auth` case (service `ssh-connection`, auth method `xyz-unknown`), but the test buffer ends exactly at the end of the auth-method name with no trailing bytes. As a result, `begin` already equals `len` before the new assignment is reached, and the test's `idx != len` assertion would still pass even if the new `begin = len;` line were removed. Because `DoReceive()` advances `inputBuffer.idx` from the handler-reported `payloadIdx` (src/internal.c:10146-10159), the PR still lacks a test proving that an unsupported auth method with extra trailing bytes is fully consumed and leaves packet decoding aligned. A regression that re-introduced an offset bug in this branch would not be caught.

**Fix:** Extend `test_DoUserAuthRequest_serviceName()` with a valid `"ssh-connection"` case that includes an unsupported auth method plus extra trailing bytes (for example, fake password-method fields), then verify `wolfSSH_TestDoUserAuthRequest()` still returns `WS_SUCCESS`, sends `MSGID_USERAUTH_FAILURE`, and advances `idx` to `len`. Without this scenario, the consumption-of-remaining-data behavior added by this PR is not under test.

**Review by aidangarske — COMMENTED** [review]
## Skoll Multi-Scan Review

**Modes:** `audit + review-security`
**Overall recommendation:** `COMMENT`
**Findings:** 1 total — 1 posted, 0 skipped
**1 finding(s) posted as inline comments** (see file-level comments below)

### Posted findings
- [Low] [audit] **Unknown-auth-method `begin = len` consumption path lacks meaningful coverage** — `src/internal.c:8402-8408`

*Review generated by [Skoll](https://github.com/wolfSSL/fenrir/tree/main/skoll)*

**Comment by dgarske** [comment]
@yosuke-wolfssl please see the feedback.

File: src/internal.c (hunk 13 above)
Comment (yosuke-wolfssl) [reply src/internal.c]: Fixed

**Comment by yosuke-wolfssl** [comment]
Hello @aidangarske ,
I'm back with Skoll now and fixed some minor issues.
Could you review this again ?

**Review by ejohnstown — APPROVED** [review]
